### PR TITLE
Preliminary Serialization Testing - Configuration and Optionals

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
* Have an automated way of ensuring we haven't accidentally broken the API _for users_, given relatively risky changes in #1999 and #2016.
* This doesn't cover everything, but is targeted at specific parts of the API that are affected by those changes.

**Implementation Description (bullets)**:
* Add deserialization tests for `AtlasDbConfig` and `TimeLockServerConfiguration` (these are intended to ensure that we cope correctly with deserializing remoting2 SSL configs)
* Add serialization tests for `LeaderElectionImpl` and `ExceptionMappers` (which exercise Optionals, as well as remoting2 SerializableErrors).

This is not comprehensive, but I think would be a good first step given the changes we've made / are about to make.

**Concerns (what feedback would you like?)**:
* We'll need to add a small change to #2016 to change the tests when Guava Optionals are removed.
* There's some duplication in how we build `ObjectMapper`s. Ideally we should unify the construction of object mappers in some kind of serialization-provider class (where reasonable).

**Where should we start reviewing?**:
There are 4 sets of tests, each of which can be reviewed independently. Take your pick.

**Priority (whenever / two weeks / yesterday)**:
Ideally before #2016 merges so we can get signal from this.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2029)

### CODE REVIEW

File: atlasdb-config/src/test/java/com/palantir/atlasdb/config/AtlasDbConfigSerializationTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Paths;
+
+import org.junit.Test;
+
+import com.palantir.atlasdb.memory.InMemoryAtlasDbConfig;
+import com.palantir.remoting2.config.ssl.SslConfiguration;
+
+public class AtlasDbConfigSerializationTest {
```
Comment (tboam) [inline atlasdb-config/src/test/java/com/palantir/atlasdb/config/AtlasDbConfigSerializationTest.java]: Is this a 'serialization' test?  I only see deserialization happening.

Serialization may not be interesting here as we don't write this out anywhere.

**Comment by tpetracca** [comment]
Per your comment above about ObjectMappers and the weirdness I've encountered the past 2 weeks I've filed #2030

**Comment by tpetracca** [comment]
@jeremyk-91 am I correct in saying that if #1999 had created a break that the tests for this PR wouldn't pass?

File: atlasdb-config/src/test/java/com/palantir/atlasdb/config/AtlasDbConfigSerializationTest.java (hunk 1 above)
Comment (jeremyk-91) [inline atlasdb-config/src/test/java/com/palantir/atlasdb/config/AtlasDbConfigSerializationTest.java]: Ah, right - there's only deserialization (I think when I wrote this I was thinking 'test the serialized form', but we don't generate on-disk configs as you point out), so it's probably clearer as `...DeserializationTest`. 

Will do this for the timelock server config one too, which is similar.

**Comment by jeremyk-91** [comment]
@tboam @tpetracca thanks for the super-prompt review!

Regarding #1999 - for the most part. As you note in the PR there are 4 main groups of changes (config, tracing, clients, jersey) - the tests exercise the config and some of the jersey stuff (ExceptionMappers), but might not catch a break in tracing or UserAgents.

I based the example configs off what we had in the docs (which were written before we bumped to remoting2). So we can be confident there wasn't a user-break in terms of config, or ExceptionMappers (important if we have new timelock talking with older clients).

**Pull request merged**